get_branch_name.py

```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/python3
+
+import os
+import subprocess
+
+def get_branch_by_hash():
+    try:
+        branch = os.environ.get('branch', "HEAD")
+        commit = os.environ.get('commit_id')
+        # commit = "f84e1cea5"
+
+        contains = subprocess.check_output(['git', 'branch', '--contains', commit])
```

Comment:
You tried this snippets on branch.

```
$ git log --oneline -n2
5a75cc8b0 (HEAD -> dev_202205_correct_branch, origin/dev_202205_correct_branch) Align codestyle, small refactor
bce1a9ed0 [202205] add step to define branch name by hash

$ git branch 
  dev_202205
* dev_202205_correct_branch

$ git checkout 5a75cc8b0
...
HEAD is now at 5a75cc8b0 Align codestyle, small refactor

$ python3
Python 3.8.10 (default, Nov 14 2022, 12:59:47) 
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import subprocess
>>> subprocess.run("git branch --contains 5a75cc8b05e73e44a0adb70250f3aef17e864b4b | grep -F '*'", shell=True, universal_newlines=True, stdout=subprocess.PIPE).stdout
'* (HEAD detached at 5a75cc8b0)\n'
>>> subprocess.run("git branch --contains 5a75cc8b0 | awk '/\*/ {print $2}'", shell=True, universal_newlines=True, stdout=subprocess.PIPE).stdout.rstrip()
'(HEAD'
```

Also I need to validate number of returned strings.

Reply:
why do we need to validate number of string?
Probably, if we will use `grep` then we can omit some code L15-L21 (including validating number of string). And just do `"(HEAD detached at " in some_str`

I am just wondering, I am not insist to use `grep`

Reply:
Let me explain. For branch (current) checkness `git rev-parse --abbrev-ref HEAD` command is used. There are 2 options of return value:
1. Branch name (if repo stands on branch)
```
$ git branch 
  dev_202205
* dev_202205_correct_branch
* 
$ git rev-parse --abbrev-ref HEAD
dev_202205_correct_branch
```

2. HEAD detached .... (if repo stands on commit)
```
$ git branch 
  dev_202205
* dev_202205_correct_branch

$ git log --oneline -n2
5a75cc8b0 (HEAD -> dev_202205_correct_branch, origin/dev_202205_correct_branch) Align codestyle, small refactor
bce1a9ed0 [202205] add step to define branch name by hash

$ git checkout 5a75cc8b0
...
HEAD is now at 5a75cc8b0 Align codestyle, small refactor

$ git branch 
* (HEAD detached at 5a75cc8b0)
  dev_202205
  dev_202205_correct_branch

$ git rev-parse --abbrev-ref HEAD
HEAD

```

As you can see from example above - commit may be in several branches as well as in none of them (commit exists but doesn't connected to any branch):
```
$ git branch 
* (HEAD detached at 5a75cc8b0)
  dev_202205
  dev_202205_correct_branch

$ echo "" > a
$ git add a
$ git commit -m "test"
[detached HEAD 4dce25f52] test
 1 file changed, 1 insertion(+)
 create mode 100644 a

$ git log --oneline -n3
4dce25f52 (HEAD) test
5a75cc8b0 (origin/dev_202205_correct_branch, dev_202205_correct_branch) Align codestyle, small refactor
bce1a9ed0 [202205] add step to define branch name by hash

$ git branch 
* (HEAD detached from 5a75cc8b0)
  dev_202205
  dev_202205_correct_branch

$ git branch --contains 4dce25f52
* (HEAD detached from 5a75cc8b0)

$ git checkout dev_202205_correct_branch
Warning: you are leaving 1 commit behind, not connected to
any of your branches:

  4dce25f52 test

If you want to keep it by creating a new branch, this may be a good time
to do so with:

 git branch <new-branch-name> 4dce25f52

Switched to branch 'dev_202205_correct_branch'
Your branch is up to date with 'origin/dev_202205_correct_branch'.

$ git branch --contains 4dce25f52
$ # empty result above
```

So that is why I need to do some validation with string number. As you can see below from comment there is one case when we are able to determine branch from commit hash:
```
        # Branch can be defined by commit just in case if "git branch --contains"
        # returns 2 strings: 1st is "(HEAD detached at {hash})" and 2nd is a branch name.
        # In other cases branch cannot be defined so return HEAD
```
And there is no way to check number of lines, validate 1st string and get second in single pipeline. There is no sense to do all validations in shell by python subprocess. In other words you need additional shell script for this.

Yep, shell scripts work faster then python scripts. BUT. The bunch of pipelines doesn't work faster than single python script because each shell command (git, awk, grep etc..) is a program. So programs in pipeline e.g. `git -> grep -> ...` doesn't reuse memory with previous return but will take another part and copy to it. Shell script- code or python code with pipelines is more complex than simple python code with one `git branch --contains <hash>` call. 

Regarding `subprocess.check_output` - I agree with you, better to use `subprocess.run`. It will do code shorter and less complex.